spark/src/test/java/org/apache/iceberg/spark/source/TestIdentityPartitionData.java

```diff
@@ -0,0 +1,165 @@
+      Types.NestedField.optional(2, "date", Types.StringType.get()),
+      Types.NestedField.optional(3, "level", Types.StringType.get()),
+      Types.NestedField.optional(4, "message", Types.StringType.get())
+  );
+
+  private static final List<LogMessage> LOGS = ImmutableList.of(
+      LogMessage.debug("2020-02-02", "debug event 1"),
+      LogMessage.info("2020-02-02", "info event 1"),
+      LogMessage.debug("2020-02-02", "debug event 2"),
+      LogMessage.info("2020-02-03", "info event 2"),
+      LogMessage.debug("2020-02-03", "debug event 3"),
+      LogMessage.info("2020-02-03", "info event 3"),
+      LogMessage.error("2020-02-03", "error event 1"),
+      LogMessage.debug("2020-02-04", "debug event 4"),
+      LogMessage.warn("2020-02-04", "warn event 1"),
+      LogMessage.debug("2020-02-04", "debug event 5")
+  );
+
+  @Rule
+  public TemporaryFolder temp = new TemporaryFolder();
+
+  private PartitionSpec spec = PartitionSpec.builderFor(LOG_SCHEMA).identity("date").identity("level").build();
+  private Table table = null;
+  private Dataset<Row> logs = null;
+
+  @Before
+  public void setupTable() throws Exception {
+    File location = temp.newFolder("logs");
+    Assert.assertTrue("Temp folder should exist", location.exists());
+
+    Map<String, String> properties = ImmutableMap.of(TableProperties.DEFAULT_FILE_FORMAT, format);
+    this.table = TABLES.create(LOG_SCHEMA, spec, properties, location.toString());
+    this.logs = spark.createDataFrame(LOGS, LogMessage.class).select("id", "date", "level", "message");
+
+    logs.write().format("iceberg").mode("append").save(location.toString());
+  }
+
+  @Test
+  public void testFullProjection() {
+    List<Row> expected = logs.orderBy("id").collectAsList();
+    List<Row> actual = spark.read().format("iceberg").load(table.location()).orderBy("id").collectAsList();
+    Assert.assertEquals("Rows should match", expected, actual);
+  }
+
+  @Test
+  public void testProjections() {
+    String[][] cases = new String[][] {
+        // individual fields
+        new String[] { "date" },
+        new String[] { "level" },
+        new String[] { "message" },
+        // field pairs
+        new String[] { "date", "message" },
+        new String[] { "level", "message" },
+        new String[] { "date", "level" },
+        // out-of-order pairs
+        new String[] { "message", "date" },
+        new String[] { "message", "level" },
+        new String[] { "level", "date" },
+        // full projection, different orderings
+        new String[] { "date", "level", "message" },
+        new String[] { "level", "date", "message" },
+        new String[] { "date", "message", "level" },
+        new String[] { "level", "message", "date" },
+        new String[] { "message", "date", "level" },
+        new String[] { "message", "level", "date" }
+    };
+
+    for (String[] ordering : cases) {
+      List<Row> expected = logs.select("id", ordering).orderBy("id").collectAsList();
```

Comment:
nit: I think the testFullProjection is a special case of testProjections , would merge them together ?

Reply:
If this test case fails, it's easier to debug and find what happened than the other case. That's why I included it.